### PR TITLE
WordPress Coding Standards

### PR DESCRIPTION
**Pull request by dshanske**
Attempts to address #32.

PHPCS identified issues with the endpoint URLs and how they were generated. Cleaned up header generation code to match similar changes in webmentions plugin and abstracted them into their own functions with filters.

### CODE REVIEW

**Comment by snarfed** [comment]
thanks for the contribution!

i'm fine with the endpoint URL changes, but we should probably drop the indentation changes. i know PHPCS complained, but it's imperfect, like any automated tool, and in this case i think better judgment says no.

i try to break lines at around 80 chars when possible, since my (and many others') editor windows aren't much wider than that. PHP's coding standards agree: http://pear.php.net/manual/en/standards.indenting.php#standards.indenting

adding a 4 char tab to every line in the file reduces the horizontal real estate we have to work with, and arguably adds no value, since the `class_exists` isn't meaningful or important to remember when writing code inside this file.

also, neither wordpress's nor PHP's coding standards say anything about this explicitly:
* https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#indentation
* http://pear.php.net/manual/en/standards.php


there's precedent for not indenting after a check in wordpress itself:
* https://github.com/WordPress/WordPress/blob/master/wp-includes/class-json.php#L2
* https://github.com/WordPress/WordPress/blob/master/wp-includes/class-simplepie.php#L2

and there's also precedent for skipping the `class_exists` check altogether in other wordpress plugins (including yours! :P):
* https://github.com/acegiak/Semantic-Linkbacks/blob/master/handler/semantic-linkbacks-microformats-handler.php
* https://github.com/dshanske/syndication-links/blob/master/includes/class-hcard-widget.php

i'd be fine with either removing the check or dropping the commit that adds the indentation from this PR. up to you!

File: micropub.php [hunk 1]
```diff
@@ -810,31 +810,44 @@ private static function check_error( $result ) {
...
+		}
+
 		/**
 	 * The micropub autodicovery meta tags
 	 */
 		public static function html_header() {
-			echo '<link rel="micropub" href="' . site_url( '?micropub=endpoint' ) . '">';
-			echo '<link rel="authorization_endpoint" href="' . MICROPUB_AUTHENTICATION_ENDPOINT . '">';
-			echo '<link rel="token_endpoint" href="' . MICROPUB_TOKEN_ENDPOINT . '">';
+			printf( '<link rel="micropub" href="%s" />' . PHP_EOL, static::get_micropub_endpoint() );
```
Comment (snarfed) [inline micropub.php]: minor nit: `link` has no content, so you can drop the closing `/`s.

File: micropub.php [hunk 2]
```diff
@@ -810,31 +810,44 @@ private static function check_error( $result ) {
 			return $result;
 		}
 
+
+		public static function get_micropub_endpoint() {
+			return apply_filters( 'micropub_endpoint', site_url( '?micropub=endpoint' ) );
```
Comment (snarfed) [inline micropub.php]: three new filters! cool! can you please add them to `README.md` and `readme.txt`, in both the _Filters and hooks_ sections and the changelogs?

**Review by snarfed — CHANGES_REQUESTED** [review]

**Comment by dshanske** [comment]
I'm going with removing the check since it would throw an error on activation or deactivation of a competing name only. And since this is the first Micropub plugin for WordPress....

File: micropub.php (hunk 1 above)
Comment (dshanske) [inline micropub.php]: I don't have a strong opinion on either side of the / debate(it seems there are differing opinions on it. I liked the changes that @pfefferle made to similar entries in the webmention plugins because they are cleaner and addressed another issue, and he had the / .

File: micropub.php (hunk 2 above)
Comment (dshanske) [inline micropub.php]: I will do that. Having the three different versions come from a function as opposed to hard-coded ensured that they'd always be set identically. And as long as there was a function required, the filter made sense.

File: micropub.php (hunk 1 above)
Comment (snarfed) [inline micropub.php]: oh huh, i didn't realize the closing `/` was allowed (optional) in empty elements: https://www.w3.org/TR/html/syntax.html#start-tags

i'm fine either way then.

File: micropub.php [hunk 3]
```diff
@@ -810,31 +810,44 @@ private static function check_error( $result ) {
...
+		}
+
+		public static function get_authentication_endpoint() {
+			return apply_filters( 'micropub_authentication_endpoint', MICROPUB_AUTHENTICATION_ENDPOINT );
+		}
+
+		public static function get_token_endpoint() {
+			return apply_filters( 'micropub_endpoint', MICROPUB_TOKEN_ENDPOINT );
```
Comment (snarfed) [inline micropub.php]: oh also, looks like this filter name needs to be `micropub_token_endpoint`

Comment (snarfed) [inline micropub.php]: ...which is a good sign that it's probably time for unit tests for these headers. :P consider adding some in this PR?

Comment (dshanske) [inline micropub.php]: Making the fixes now.

File: readme.txt [hunk 4]
```diff
@@ -90,9 +96,9 @@ These configuration options can be enabled by adding them to your wp-config.php
 * `define('MICROPUB_LOCAL_AUTH', '1')` - Bypasses Micropub authentication in
 favor of WordPress authentication
 * `define('MICROPUB_AUTHENTICATION_ENDPOINT', 'https://indieauth.com/auth')` -
-Define a custom authentication endpoint
+Define a custom authentication endpoint (can also be overridden by filter)
```
Comment (snarfed) [inline readme.txt]: minor nit: consider mentioning the name of the filter :P (also below)

File: README.md [hunk 5]
```diff
@@ -25,10 +25,15 @@ sideloaded from URLs.
 
 #### Filters and hooks
 
-Adds one filter: `before_micropub( $input )`
+Adds several filter: `before_micropub( $input )`
```
Comment (snarfed) [inline README.md]: minor nit, typo: filter*s*

**Comment by snarfed** [comment]
thanks for the updates! some of the other style changes seem surprising too, but i'll check the wordpress and php standards and bite my tongue if they're all explicitly recommended. :P

otherwise, maybe consider a couple things:
* squash the two style commits? the indentation changes and then rollbackas are both pretty invasive, which causes problems when git annotating and following code changes across commits. this might help: https://help.github.com/articles/using-git-rebase-on-the-command-line/
* unit tests for the headers and filters :P

File: README.md [hunk 6]
```diff
@@ -45,6 +50,8 @@ Arguments:
   arguments passed to wp_insert_post or wp_update_post. For deletes and
   undeletes, args['ID'] contains the post id to be (un)deleted. Null for queries.
 
+`$default`: The default endpoint location used by the plugin. The Authentication and Token Endpoints default to IndieAuth, the Micropub endpoint to the /?micropub=endpoint
```
Comment (snarfed) [inline README.md]: minor nits:
* please link IndieAuth (ie `[IndieAuth](https://indieauth.com/)`)
* _the /?micropub=endpoint_ => _\`/?micropub=endpoint\`_.

also in readme.txt.

**Comment by dshanske** [comment]
I'll squash them. I was in the process of figuring out how to test a header.

File: micropub.php [hunk 7]
```diff
@@ -122,7 +120,7 @@ public static function parse_query( $wp ) {
 		static::load_input();
 		if ( WP_DEBUG ) {
 			error_log( 'Micropub Data: ' . serialize( $_GET ) . ' ' .
-					   serialize( static::$input ) );
+			serialize( static::$input ) );
```
Comment (snarfed) [inline micropub.php]: please also revert these changes that remove indentation from line continuations. the indentation is needed for readability, and both wordpress's and PHP's code standards agree:

https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#indentation
http://pear.php.net/manual/en/standards.funcalls.php#standards.funcalls.splitfunccall

again, not sure what PHPCS is smoking. :P

File: micropub.php [hunk 8]
```diff
@@ -155,53 +153,55 @@ private static function authorize() {
...
+			$code, 'invalid access token: ' . $body );
 		} elseif ( ! in_array( 'post',  $scopes ) && ! in_array( 'create',  $scopes ) ) {
 			static::handle_authorize_error(
-				401, 'access token is missing post or create scope; got `' . $params['scope'] . '`' );
+			401, 'access token is missing post or create scope; got `' . $params['scope'] . '`' );
 		}
 
-		parse_str( $body, $resp );
-		$me = untrailingslashit( $resp['me'] );
+			parse_str( $body, $resp );
```
Comment (snarfed) [inline micropub.php]: these are wrong. also below. please revert.

File: micropub.php [hunk 9]
```diff
@@ -212,23 +212,23 @@ public static function post_handler( $user_id ) {
...
 			if ( $user_id && ! user_can( $user_id, 'publish_posts' ) ) {
 				static::error( 403, 'user id ' . $user_id . ' cannot publish posts' );
 			}
 			$args = static::create( $user_id );
 			$status = 201;
 			static::header( 'Location', get_permalink( $args['ID'] ) );
 
-		// update
-		} elseif ( $action == 'update' || ! $action ) {
+			// update
```
Comment (snarfed) [inline micropub.php]: same, revert, and below

File: micropub.php [hunk 10]
```diff
@@ -805,56 +805,65 @@ private static function check_error( $result ) {
...
-				} elseif ( $key == 'h' ) {
-					static::$input['type'] = array( 'h-' . $val );
-				} elseif ( $key == 'access_token' ) {
-					continue;
-				} else {
-					if ( ! isset( static::$input['properties'] ) ) {
-						static::$input['properties'] = array();
-					}
-					static::$input['properties'][ $key ] =
+		public static function jrd_links( $array ) {
```
Comment (snarfed) [inline micropub.php]: please squash these out entirely too. thanks!

File: micropub.php [hunk 11]
```diff
@@ -696,14 +699,14 @@ public static function store_geodata( $args ) {
 				$coords = explode( ',', $geo[0] );
 				$args['meta_input']['geo_latitude'] = trim( $coords[0] );
 				$args['meta_input']['geo_longitude'] = trim( $coords[1] );
-			} elseif ( substr( $location, 0, 4 ) != 'http' ) {
+			} elseif ( substr( $location, 0, 4 ) !== 'http' ) {
```
Comment (snarfed) [inline micropub.php]: also switch order to make this yoda style like the others?

**Review by snarfed — CHANGES_REQUESTED** [review]
thanks for squashing!

i looked at the rest of the style changes, and honestly, PHPCS seems pretty flawed at best. switching to "yoda style" conditionals and the few missing spaces inside parens are good, but switching to `===` and `!==` doesn't really matter in any of our cases, and all of the indentation changes were wrong. ah well!

**Comment by snarfed** [comment]
honestly, it might be easier to start over and just do the endpoint and yoda style conditional changes, instead of trying to extract them out of these commits. up to you!

**Comment by dshanske** [comment]
I thought it was easier to address the coding standard thing because it would move everything around. I think it doesn't necessarily help much. I may just leave that part for now, close this, and focus on adding functionality in a separate PR.
